src/main/kotlin/org/jetbrains/research/libsl/nodes/references/TypeReference.kt

```diff
@@ -3,55 +3,152 @@ package org.jetbrains.research.libsl.nodes.references
-    open var typeBound: GenericTypeBound = GenericTypeBound.EMPTY,
-    open val genericReferences: MutableList<TypeReference> = mutableListOf(),
+interface TypeReference : LslReference<Type, TypeReference> {
     override val context: LslContextBase
-) : LslReference<Type, TypeReference> {
+}
+
+data class UnionExpressionTypeReference(
+    val left: TypeReference,
+    val right: TypeReference,
+    override val context: LslContextBase
+) : TypeReference {
+    override fun resolve(): Type? {
+        this.left.resolve()
+        this.right.resolve()
+        // We don't have type for UnionExpression
+        return null
+    }
+
+    override fun isReferenceMatchWithNode(node: Type): Boolean {
+        // we don't have Type for UnionExpressionTypeReference
+        return false
+    }
+
+    override fun isSameReference(other: TypeReference): Boolean {
+        // This is right realization ??
+        if (other is UnionExpressionTypeReference) {
+            this.left.isSameReference(other.left) && this.right.isSameReference(other.right)
+        }
+        return false
+    }
+
+    fun getName(): String {
+        return this.left.getName() + " | " + this.right.getName()
+    }
+
+}
+
+data class IntersectionExpressionTypeReference(
+    val left: TypeReference,
+    val right: TypeReference,
+    override val context: LslContextBase
+) : TypeReference {
+    override fun resolve(): Type? {
+        this.left.resolve()
+        this.right.resolve()
+        // We don't have type for IntersectionExpression
+        return null
+    }
+
+    override fun isReferenceMatchWithNode(node: Type): Boolean {
+        // we don't have Type for IntersectionExpressionTypeReference
+        return false
+    }
+
+    override fun isSameReference(other: TypeReference): Boolean {
+        // This is right realization ??
+        if (other is IntersectionExpressionTypeReference) {
+            this.left.isSameReference(other.left) && this.right.isSameReference(other.right)
+        }
+        return false
+    }
+
+    fun getName(): String {
+        return this.left.getName() + " & " + this.right.getName()
+    }
+}
+
+data class LiteralTypeReference(
+    val value: String,
+    override val context: LslContextBase
+) : TypeReference {
     override fun resolve(): Type? {
-        return resolveArrayType() ?: resolveListType() ?: resolveMapType() ?: resolveNullType()
+        return resolveLiteralType()
+    }
+
+    private fun resolveLiteralType(): Type? {
+        if (value.startsWith("\""))
+            return StringType(context)
+        else if (value.startsWith("\'"))
+            return CharType(context)
+        else if ((Character.isDigit(value.first()) && value.contains(",")) || (value.startsWith("-") && value.contains(",")))
+            return Float64Type(context)
+        else if (Character.isDigit(value.first()) || value.startsWith("-"))
+            return Int64Type(context)
+        return null
+    }
```

Comment:
Can `Type` information be resolved at construction time?
(i.e. then making a call to the constructor of this reference)

Reply:
@dog-m Save field "Type" in LiteralTypeReference ?

Reply:
Fixed;